sdk/include/opentelemetry/sdk/http/http_client.h

```diff
@@ -0,0 +1,153 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+#include "opentelemetry/http/http_client.h"
+#include "opentelemetry/version.h"
+
+#include <algorithm>
+#include <cstring>
+#include <functional>
+#include <map>
+#include <memory>
+#include <vector>
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace http
+{
+namespace http_api = opentelemetry::http;
+
+static std::function<bool(const std::string &, const std::string &)> CaseInsensitiveComparator =
+    [](const std::string &s1, const std::string &s2) -> bool {
+  std::string str1(s1.length(), ' ');
+  std::string str2(s2.length(), ' ');
+  auto lowerCase = [](char c) -> char { return tolower(c); };
+  std::transform(s1.begin(), s1.end(), str1.begin(), lowerCase);
+  std::transform(s2.begin(), s2.end(), str2.begin(), lowerCase);
+  return str1 < str2;
+};
+
+// HttpHeaders implementation
+class HttpHeaders : http_api::HttpHeaders,
+                    std::multimap<std::string, std::string, decltype(CaseInsensitiveComparator)>
+{
+public:
+  virtual void set(nostd::string_view const &name, nostd::string_view const &value) override
+  {
+    auto range = equal_range(std::string(name.data()));
+    auto hint  = erase(range.first, range.second);
+    insert(hint, std::make_pair(name, value));
+  }
+
+  virtual void add(nostd::string_view const &name, nostd::string_view const &value) override
+  {
+    insert(std::make_pair(name, value));
+  }
+
+  virtual nostd::string_view const &get(nostd::string_view const &name) const override
```

Comment:
Will we support multiple values for a single header?

Reply:
Maybe we should make it a case-insensitive multimap, but with about the same interface as `std::multimap`

Reply:
I think:
1. We need to support multiple headers with the same name.
2. We need to preserve the upper/lower case.
3. For set/get scenario, it _might_ make sense to be case-insensitive.

Reply:
It's be _more convenient_ to have a case-insensitive getter :) ... I also kinda agree we should store preserving case, even if HTTP spec suggests otherwise.

Reply:
So to summarise,  will preserve case while storing keys in multimap, even though storage order would be case-insensitive ( i.e, custom Comparison operator). And getter's would be case-insensitive for keys.